File: Assembler Java/src/Symbol_Table.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.util.ArrayList;
import java.util.Map;
import java.util.LinkedHashMap;

public class Symbol_Table {
    public static final ArrayList<String> FileLines = No_White_Space.Remove_White_Space();
    public static final LinkedHashMap<String, String> SymbolTable = new LinkedHashMap<>();

    static {
        SymbolTable.put("R0", "0");
        SymbolTable.put("R1", "1");
        SymbolTable.put("R2", "2");
        SymbolTable.put("R3", "3");
        SymbolTable.put("R4", "4");
        SymbolTable.put("R5", "5");
        SymbolTable.put("R6", "6");
        SymbolTable.put("R7", "7");
        SymbolTable.put("R8", "8");
        SymbolTable.put("R9", "9");
        SymbolTable.put("R10", "10");
        SymbolTable.put("R11", "11");
        SymbolTable.put("R12", "12");
        SymbolTable.put("R13", "13");
        SymbolTable.put("R14", "14");
        SymbolTable.put("R15", "15");
        SymbolTable.put("SP", "0");
        SymbolTable.put("LCL", "1");
        SymbolTable.put("ARG", "2");
        SymbolTable.put("THIS", "3");
        SymbolTable.put("THAT", "4");
        SymbolTable.put("SCREEN", "16384");
        SymbolTable.put("KBD", "24576");
    }

    public static void main(String[] args) {
        try {
            int bitCount = 15;
            int lineNum = 0;
            for (String line : FileLines) {
                if (line.startsWith("(")) {
                    String symbol = line.substring(1, line.length() - 1);
                    SymbolTable.put(symbol, Integer.toString(lineNum));
                }
                else{
                    lineNum++;
                }
            }
            for (String line : FileLines) {
                if (line.startsWith("@")) {
                    String symbol = line.substring(1);
                    if (SymbolTable.containsKey(symbol)) {
                        // No action needed
                    }
                    else if (symbol.matches("\\d+")) {
                        // No action needed
                    }else {
                        bitCount++;
                        SymbolTable.put(symbol, Integer.toString(bitCount));
                    }
                }
            }

            BufferedWriter myWriter = new BufferedWriter(new FileWriter("SymbolTable.asm"));
            myWriter.write("Symbols \t Values\n");
            for (Map.Entry<String, String> entry : SymbolTable.entrySet()) {
                System.out.println(entry.getKey() + " " + entry.getValue());
                myWriter.write(String.format("%-15s %-15s%n", entry.getKey(), entry.getValue()));
            }
            myWriter.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: Assembler Python/Symbol_Table.py
# This is the code for making a symbol table which holds all the symbols and variables in asm file along with their values.
# This is also part of the white.py code as an asm file with no white spaces is required for this code.
# predefined symbols are added to a dictionary.
symbol_table = {'R0':'0','R1':'1','R2':'2','R3':'3','R4':'4','R5':'5',
'R6':'6','R7':'7','R8':'8','R9':'9','R10':'10','R11':'11',
'R12':'12','R13':'13','R14':'14','R15':'15','SP':'0',
'LCL':'1','ARG':'2','THIS':'3','THAT':'4','SCREEN':'16384',
'KBD':'24576'}

# I have named the file white.asm in the previous code so continuing with the same name here :)
with open("white.asm") as white:
    with open("symboltable.asm","w") as out:
        file = white.readlines()
        bit_count = 15
        for line in file :
            # Checking whether it is an A instruction
            if line[0]==('@'):
            	# Ignoring all predefined symbols and newly added symbols
                if line[1:-1] in symbol_table.keys():
                    pass
                # Ignoring all numbers
                elif line[1:-1].isdigit() :
                    pass
                # Gives value for labels according to line number
                elif line[1].isupper():
                    line_num=0 # Line number
                    for temp in file:
                        if temp[0] =='(':
                            if line[1:-1] == temp[1:-2]:
                                symbol_table[line[1:-1]] = str(line_num)
                        else:
                            line_num+=1
                else:
                    bit_count+=1
                    symbol_table[line[1:-1]]=bit_count
        out.write("Symbols \t Values\n")
        for i,j in symbol_table.items():
            out.write("{:<15} {:<15}\n".format(i,j))